src/python/pants/engine/legacy/structs.py

```diff
@@ -308,38 +298,39 @@ class PythonRequirementLibraryAdaptor(TargetAdaptor): pass
 
 
 class PantsPluginAdaptor(PythonTargetAdaptor):
-  def get_sources(self):
+  def get_sources(self) -> "GlobsWithConjunction":
     return GlobsWithConjunction.for_literal_files(['register.py'], self.address.spec_path)
 
 
 class BaseGlobs(Locatable, metaclass=ABCMeta):
   """An adaptor class to allow BUILD file parsing from ContextAwareObjectFactories."""
 
   @staticmethod
-  def from_sources_field(sources, spec_path):
-    """Return a BaseGlobs for the given sources field.
-
-    `sources` may be None, a list/tuple/set, a string or a BaseGlobs instance.
-    """
+  def from_sources_field(
+    sources: Union[None, str, Iterable[str], "BaseGlobs"], spec_path: str,
+  ) -> "BaseGlobs":
+    """Return a BaseGlobs for the given sources field."""
     if sources is None:
       return Files(spec_path=spec_path)
-    elif isinstance(sources, BaseGlobs):
+    if isinstance(sources, BaseGlobs):
       return sources
-    elif isinstance(sources, str):
+    if isinstance(sources, str):
       return Files(sources, spec_path=spec_path)
-    elif isinstance(sources, (MutableSet, MutableSequence, tuple)) and \
-         all(isinstance(s, str) for s in sources):
+    if (
+      isinstance(sources, (MutableSet, MutableSequence, tuple))
+      and all(isinstance(s, str) for s in sources)
+    ):
       return Files(*sources, spec_path=spec_path)
-    else:
-      raise ValueError('Expected either a glob or list of literal sources: got: {}'.format(sources))
+    raise ValueError(f'Expected either a glob or list of literal sources. Got: {sources}')
 
   @staticmethod
-  def _filespec_for_exclude(raw_exclude, spec_path):
+  def _filespec_for_exclude(raw_exclude: List[str], spec_path: str) -> wrapped_globs._GlobsDict:
     if isinstance(raw_exclude, str):
-      raise ValueError('Excludes of type `{}` are not supported: got "{}"'
-                       .format(type(raw_exclude).__name__, raw_exclude))
+      raise ValueError(
+        f'Excludes should be a list of strings. Got: {raw_exclude!r}'
+      )
```

Comment:
This error message would only ever trigger when `raw_exclude` is a `str`, so there's no need to run `type(raw_exclude).__name__`. Instead, I reworded it to suggest what we expect.

I left off in the suggestion that you can technically use a `glob` in the `exclude` field, as that is soon to be deprecated.